File: 2022/01/22/rust/src/maximum_subarray.rs
// Program: https://leetcode.com/problems/maximum-subarray/
// Author: Andrew Jarombek
// Date: 1/22/2022

pub fn max_sub_array(nums: Vec<i32>) -> i32 {
    let mut max = i32::MIN;
    let mut current = 0;

    for num in nums {
        max = std::cmp::max(current + num, max);
        current = std::cmp::max(0, current + num);
    }

    return max;
}

File: 2022/02/10/maximumSubarray.swift
// https://leetcode.com/problems/maximum-subarray/
// Author: Andrew Jarombek
// Date: 2/10/2022

func maxSubArray(_ nums: [Int]) -> Int {
    var largest = nums[0]
    var current = 0

    for i in 0..<nums.count {
        current += nums[i]

        if current > largest {
            largest = current
        }

        if current < 0 {
            current = 0
        }
    }

    return largest
}